hackers delight

AnotherAlexHere opened the pull request.

AnotherAlexHere committed:
chore: def to lemma

AnotherAlexHere committed:
Merge github.com:opencompl/lean-mlir into def_to_lemma

AnotherAlexHere committed:
chore: stated theorems

AnotherAlexHere committed:
chore: fixed formatting

tobiasgrosser commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,101 @@
+  apply BitVec.eq_of_toNat_eq
+  sorry
+
+
```
Do not use two empty lines in sequence.

AnotherAlexHere replied:
Will fix this in next commit.

tobiasgrosser replied:
And this line to disappear.

tobiasgrosser commented on SSA/Projects/InstCombine/ForLean.lean:
```diff
@@ -37,19 +37,19 @@ lemma two_pow_pred_mod_eq_two_pred (h : w > 0): 2 ^ (w - 1) % 2 ^ w = 2 ^ (w - 1
 
-def ushr_or_distrib (a b c : BitVec w) :
+lemma ushr_or_distrib (a b c : BitVec w) :
     (a ||| b) >>> c = (a >>> c) ||| (b >>> c) := by
```
These changes are unrelated and should not be included here.

AnotherAlexHere replied:
Yes, you are right.

tobiasgrosser replied:
I want these changes to disappear. 😄

tobiasgrosser replied:
Now these are back again. The last revert was not really needed, I think.

tobiasgrosser replied:
Try to work with @AtticusKuhn to clean up this PR. You may need some help with checking how to make this clean and ready.

tobiasgrosser commented:
@AnotherAlexHere, please wait for at least one other person to review your changes.

tobiasgrosser commented:
@AnotherAlexHere, if you could address my comments, that would be appreciated.

AnotherAlexHere committed:
Merge branch 'main' of github.com:opencompl/lean-mlir

AnotherAlexHere committed:
chore: comments

AnotherAlexHere committed:
Merge branch 'HackersDelight1'

AnotherAlexHere committed:
feat: some new proofs

AnotherAlexHere committed:
Merge branch 'HackersDelightTemp' into HackersDelight1

tobiasgrosser commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,110 @@
+theorem neg_add_eq_neg_sub {x: BitVec w } :
+    -(x + y) = -x - y := by
+    ring
+
```
```suggestion
  ring

```

Tactics are indentented by two whitespace only.

tobiasgrosser commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,110 @@
+theorem not_eq_neg_sub_one {x : BitVec w} :
+    ~~~ x = -x - 1 := by
+  apply BitVec.eq_sub_iff_add_eq.mpr
+  rw[BitVec.neg_eq_not_add]
```
```suggestion
  rw [BitVec.neg_eq_not_add]
```

Whitespace between rw and `[`.

tobiasgrosser commented:
@AnotherAlexHere, I am not sure we want to try to 'proof' these statements. I would suggest keeping the proofs where they go through and just leaving sorry everywhere else.

tobiasgrosser commented:
The one thing that is important is that we actually build this file. Maybe @AtticusKuhn can help us to get it hooked to the build system?

AnotherAlexHere committed:
Apply suggestions from code review

Co-authored-by: Tobias Grosser <tobias@grosser.es>

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,110 @@
+import Init.Data.BitVec.Basic
```
Rename this file to `HackersDelight.lean` so we get lean syntax highlighting

AnotherAlexHere committed:
chore: whitespaces

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,110 @@
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
+  ring
+
+theorem or_eq_not_or_sub_not {x y : BitVec w} :
```
```suggestion
theorem or_eq_and_not_sub {x y : BitVec w} :
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight:
```diff
@@ -0,0 +1,110 @@
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
```
Put a space after rw
```suggestion
  rw [not_eq_neg_sub_one]
```

AnotherAlexHere committed:
chore: added .lean ext

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem neg_eq_neg_not_one {x : BitVec w} :
+    -x = ~~~(x - 1) := by
+  rw[not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one]
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem not_not {x : BitVec w} :
+    ~~~ ~~~ x = x := by
+  rw[not_eq_neg_sub_one,not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one,not_eq_neg_sub_one]
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem neg_not_eq_add_one {x : BitVec w} :
+    - ~~~ x = x + 1 := by
+  rw[BitVec.neg_eq_not_add, not_not]
```
```suggestion
  rw [BitVec.neg_eq_not_add, not_not]
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem not_neg_eq_sub_one {x : BitVec w} :
+    ~~~ (-x) = x - 1 := by
+  rw[not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one]
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem add_eq_sub_not_sub_one {x y : BitVec w} :
+    x + y = x - ~~~ y - 1 := by
+  rw[not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one]
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
+  ring
+
+theorem or_eq_not_or_sub_not {x y : BitVec w} :
```
```suggestion
theorem or_eq_and_not_add {x y : BitVec w} :
```

AtticusKuhn commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem not_sub_eq_sub_sub_one {x y : BitVec w} :
+    ~~~ (x - y) = y - x - 1 := by
+  rw[not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one]
```

AnotherAlexHere committed:
chore: Atticus review changes

Co-authored-by: Atticus Kuhn <52258164+AtticusKuhn@users.noreply.github.com>

AnotherAlexHere commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem sub_eq_add_not_add_one {x y : BitVec w} :
+    x - y = x + ~~~ y + 1 := by
+  rw[not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one]
```

AnotherAlexHere commented on SSA/Projects/InstCombine/HackersDelight.lean:
```diff
@@ -1,14 +1,110 @@
+
+theorem not_sub_eq_not_add {x y : BitVec w} :
+    ~~~ (x - y) = ~~~ x + y := by
+  rw[not_eq_neg_sub_one, not_eq_neg_sub_one]
```
```suggestion
  rw [not_eq_neg_sub_one, not_eq_neg_sub_one]
```

AnotherAlexHere committed:
chore: more whitespaces

Co-authored-by: Atticus Kuhn <52258164+AtticusKuhn@users.noreply.github.com>

Commit:
Add Github workflow to check Hacker's Delight

added a workflow to the GitHub CI which will run and check Hacker's
Delight on each branch push

Commit:
remove excessive emojis

AtticusKuhn commented:
> The one thing that is important is that we actually build this file. Maybe @AtticusKuhn can help us to get it hooked to the build system?

I added the following job to the Github CI


```yaml
      - name: Compile Hacker's Delight Theorems 🧮
        run: |
          lake -R exe cache get # download cache of mathlib docs.
          lake -R build SSA.Projects.InstCombine.HackersDelight # compile and check the Hacker's Delight theorems
```

Hopefully, this should work.

tobiasgrosser commented on .github/workflows/ci.yml:
```diff
@@ -36,6 +36,12 @@ jobs:
+          lake -R build SSA.Projects.InstCombine.HackersDelight # compile and check the Hacker's Delight theorems
+
+
       - name: Compile Alive Examples
```
```suggestion

      - name: Compile Alive Examples
```

tobiasgrosser committed:
Update .github/workflows/ci.yml
